packages/backend/src/services/CatalogService/CatalogService.ts

```diff
@@ -235,6 +237,100 @@ export class CatalogService<
         );
     }
 
+    async indexCatalog(
+        projectUuid: string,
+        explores: (Explore | ExploreError)[],
+    ) {
+        const exploresWithCachedExploreUuid =
+            await this.projectModel.getCachedExploresWithUuid(
+                projectUuid,
+                explores,
+            );
+
+        return this.catalogModel.indexCatalog(
+            projectUuid,
+            exploresWithCachedExploreUuid,
+        );
+    }
+
+    /**
+     * Migrates catalog item tags from the previous catalog to the new catalog
+     * This is a best effort migration, if the catalog item is not found in the new catalog
+     * it will not be migrated
+     *
+     * For the item to be found it needs to:
+     * - have the same name
+     * - have the same explore base table
+     * - have the same field type
+     * - have the same type
+     * - be in the same project
+     * @param projectUuid - project uuid
+     * @param prevCatalogItemsWithTags - catalog items with tags before the catalog being re-indexed
+     * @returns
+     */
+    async migrateCatalogItemTags(
+        projectUuid: string,
+        prevCatalogItemsWithTags: CatalogItemWithTagUuids[],
+    ) {
+        const currentCatalogItems =
+            await this.catalogModel.getCatalogItemsAndTags(projectUuid);
```

Comment:
wonder if we can get just the catalog items that have tags applied to them. Do you think this would benefit us in terms of performance?

Reply:
so some early filtering could make the map faster and we wouldn't need to do the check inside the `flatMap`

Reply:
Done in: [chore: add option to return only tagged items](https://github.com/lightdash/lightdash/pull/12294/commits/7757b861d72cd6f054633119cf7eb6869062a67c)